lib/location/method_gnss.c

```diff
@@ -0,0 +1,835 @@
+		break;
+
+	case LOC_ACCURACY_NORMAL:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+		break;
+
+	case LOC_ACCURACY_HIGH:
+		use_case = NRF_MODEM_GNSS_USE_CASE_MULTIPLE_HOT_START;
+		err |= nrf_modem_gnss_use_case_set(use_case);
+
+		/* In high accuracy mode, use the configured fix count. */
+		fixes_remaining = gnss_config.num_consecutive_fixes;
+		break;
+	}
+
+	if (err) {
+		LOG_ERR("Failed to configure GNSS");
+		loc_core_event_cb_error();
+		running = false;
+		return;
+	}
+
+	err = nrf_modem_gnss_start();
+	if (err) {
+		LOG_ERR("Failed to start GNSS");
+		loc_core_event_cb_error();
+		running = false;
+		return;
+	}
+
+	loc_core_timer_start(gnss_config.timeout);
+}
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+static void method_gnss_date_time_event_handler(const struct date_time_evt *evt)
+{
+	switch (evt->type) {
+	case DATE_TIME_OBTAINED_MODEM:
+		LOG_DBG("DATE_TIME_OBTAINED_MODEM");
+		break;
+	case DATE_TIME_OBTAINED_NTP:
+		LOG_DBG("DATE_TIME_OBTAINED_NTP");
+		break;
+	case DATE_TIME_OBTAINED_EXT:
+		LOG_DBG("DATE_TIME_OBTAINED_EXT");
+		break;
+	case DATE_TIME_NOT_OBTAINED:
+		LOG_DBG("DATE_TIME_NOT_OBTAINED");
+		break;
+	default:
+		break;
+	}
+}
+#endif
+
+int method_gnss_location_get(const struct loc_method_config *config)
+{
+	const struct loc_gnss_config gnss_config = config->gnss;
+	int err;
+
+	if (running) {
+		LOG_ERR("Previous operation ongoing.");
+		return -EBUSY;
+	}
+
+	/* GNSS event handler is already set once in method_gnss_init(). If no other thread is
+	 * using GNSS, setting it again is not needed.
+	 */
+	err = nrf_modem_gnss_event_handler_set(method_gnss_event_handler);
+	if (err) {
+		LOG_ERR("Failed to set GNSS event handler, error %d", err);
+		return -err;
+	}
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+	if (!date_time_is_valid()) {
+		date_time_update_async(method_gnss_date_time_event_handler);
+	}
+#endif
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+	/* P-GPS is only initialized here because initialization may trigger P-GPS data request
+	 * which would fail if the device is not registered to a network.
+	 */
+	static bool initialized;
+
+	if (!initialized) {
+		struct nrf_cloud_pgps_init_param param = {
+			.event_handler = method_gnss_pgps_handler,
+			.storage_base = PM_MCUBOOT_SECONDARY_ADDRESS,
+			.storage_size = PM_MCUBOOT_SECONDARY_SIZE};
+
+		err = nrf_cloud_pgps_init(&param);
+		if (err) {
+			LOG_ERR("Error from PGPS init: %d", err);
+		} else {
+			initialized = true;
+		}
+	}
+#endif
+#if defined(CONFIG_NRF_CLOUD_AGPS) || defined(CONFIG_NRF_CLOUD_PGPS)
+	/* Start and stop GNSS just to see if A-GPS data is needed
+	 * (triggers event NRF_MODEM_GNSS_EVT_AGPS_REQ)
+	 */
+	nrf_modem_gnss_start();
+	nrf_modem_gnss_stop();
+#endif
+	k_work_init(&method_gnss_start_work.work_item, method_gnss_positioning_work_fn);
+	method_gnss_start_work.gnss_config = gnss_config;
+	k_work_submit_to_queue(loc_core_work_queue_get(), &method_gnss_start_work.work_item);
```

Comment:
What happens if method_gnss_location_get() is called when LTE is already in PSM, but the GNSS start/stop triggers an A-GPS data request? The A-GPS data request does cause the RRC connection to be established, but I'm afraid it might not happen quickly enough and GNSS may be started immediately. Because we don't yet have an API to query the assistance data need from GNSS before starting it, should the work which starts GNSS be delayed a bit to make sure the possible A-GPS data request gets handled before GNSS is started?

Reply:
That's a valid assumption. However, I have not seen this happen and I double checked just now. In the light of these facts, what do you propose we do? If we add a delay, I would want it to be a well justifiable number (i.e. as small as possible) because any delay here kind of makes the the TTFF longer.

Reply:
I'd ask here what's the penalty if we don't add delay? We start GNSS which doesn't get runtime, and its timer runs. We get A-GPS data and with the help of assistance we get a fix in a more reasonable time. If timeout is rather tight, we would timeout before getting the fix but I'd expect timeouts to be long enough for this.
Given above, and what Tuomas says about longer TTFF and that this has never happened in his experiments (doesn't mean it couldn't happen), we really need to have strong reasoning to add delay.

Reply:
As discussed, GNSS first sends the A-GPS data request notification and then responds to the start request. Probably because of this the library receives the A-GPS data request early enough. No changes needed, resolving.